iocuddle-sgx/src/lib.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2019 Red Hat, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#![deny(clippy::all)]
+#![allow(clippy::identity_op)]
+
+use std::marker::PhantomData;
+
+use bitflags::bitflags;
+use iocuddle::*;
+use sgx_types::{page, secs, sig};
+
+const SGX: Group = Group::new(0xA4);
+pub const ENCLAVE_CREATE: Ioctl<Write, &Create> = unsafe { SGX.write(0x00) };
+pub const ENCLAVE_ADD_PAGES: Ioctl<WriteRead, &AddPages> = unsafe { SGX.write_read(0x01) };
+pub const ENCLAVE_INIT: Ioctl<Write, &Init> = unsafe { SGX.write(0x02) };
+pub const ENCLAVE_SET_ATTRIBUTE: Ioctl<Write, &SetAttribute> = unsafe { SGX.write(0x03) };
+
+bitflags! {
+    pub struct Flags: u64 {
+        const MEASURE = 1 << 0;
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct Create<'a>(u64, PhantomData<&'a ()>);
+
+impl<'a> Create<'a> {
+    pub fn new(secs: &'a secs::Secs) -> Self {
+        Create(secs as *const _ as _, PhantomData)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct AddPages<'a> {
+    src: u64,
+    offset: u64,
+    length: u64,
+    secinfo: u64,
+    flags: Flags,
+    count: u64,
+    phantom: PhantomData<&'a ()>,
+}
+
+impl<'a> AddPages<'a> {
+    pub fn new(data: &'a [u8], offset: usize, secinfo: &'a page::SecInfo, flags: Flags) -> Self {
+        Self {
+            src: data.as_ptr() as _,
+            offset: offset as _,
+            length: data.len() as _,
+            secinfo: secinfo as *const _ as _,
+            flags,
+            count: 0,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn count(&self) -> u64 {
+        self.count
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct Init<'a>(u64, PhantomData<&'a ()>);
+
+impl<'a> Init<'a> {
+    pub fn new(sig: &'a sig::Signature) -> Self {
+        Init(sig as *const _ as _, PhantomData)
+    }
+}
+
+#[repr(C)]
+#[derive(Debug)]
+pub struct SetAttribute<'a>(u64, PhantomData<&'a ()>);
+
+impl<'a> SetAttribute<'a> {
+    pub fn new(fd: &'a impl std::os::unix::io::AsRawFd) -> Self {
+        SetAttribute(fd.as_raw_fd() as _, PhantomData)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use std::convert::TryFrom;
+    use std::fs::File;
+
+    use openssl::{bn, pkey, rsa};
+    use rstest::*;
+    use sgx_crypto::{Hasher, Signer};
+    use sgx_types::{attr, misc, page};
+
+    #[repr(C, align(4096))]
```

Comment:
Yes and no. Technically yes. Practically, almost all platforms use 4k. Specifically, SGX is only on Intel which only supports 4k pages.

Reply:
IIUC then PDF page 172 of https://www.amd.com/system/files/TechDocs/24593.pdf says that a page could be 4k, 2m or 1g. Doesn't this apply here as well?

Reply:
It might. Though I'm not sure I've ever seen Linux with a page size other than 4096. There's also the question about how to make alignment variable. I don't think we can pass a constant to this attribute. We could make a compiler plugin that emits proper page alignment on a per-target basis. But I'm not even sure where to get this information. It isn't contained in the Rust `target-spec-json`. And on some systems it is runtime configurable. So I don't see any good option besides hard-coding a value here.

@haraldh Any thoughts?

Reply:
untested idea
```rust
#[cfg_attr(page_4k, repr(C, align(4096))]
struct Page([u8; 4096]);
#[cfg_attr(page_2M, repr(C, align(2*1024*1024))]
struct Page([u8; 2*1024*1024]);
#[cfg_attr(page_1G, repr(C, align(1024*1024*1024))]
struct Page([u8; 1024*1024*1024]);
```

Reply:
@steveeJ Is this a problem we can solve in the future? In particular, I think the right solution is adding `#[repr(C, align(page))]` to Rust.

Reply:
Unfortunately, page size is not a compile time constant on some platforms (like Arm). I talked through the options with some compiler people and they agreed that having a magic number here for such architecture dependent code (Intel SGX on Linux) is probably the easiest solution.

Reply:
Well, then let's settle for easiest :smiley: